packages/tree-view/src/ux-tree-view/ux-tree-view.ts

```diff
@@ -0,0 +1,144 @@
+
+  constructor(private element: HTMLElement, container: Container, private styleEngine: StyleEngine,
+    public defaultConfiguration: UxDefaultTreeViewConfiguration, private taskQueue: TaskQueue) {
+    if (this.defaultConfiguration.theme) {
+      this.theme = this.defaultConfiguration.theme;
+    }
+    const parent = container.parent?.get(Optional.of(UxTreeView));
+    const isRoot = !parent;
+    if (isRoot) {
+      const nodeTemplateId = this.element.getAttribute('data-template-id');
+      if (nodeTemplateId && templateLookup[nodeTemplateId]) {
+        const nodeTemplate = templateLookup[nodeTemplateId];
+        const nodeViewFactory = container.get(ViewCompiler).compile(`<template>${nodeTemplate}</template>`, container.get(ViewResources));
+        this.nodeViewFactory = nodeViewFactory;
+      } else {
+        // create a default <tree-node/> factory
+        this.nodeViewFactory = container.get(ViewCompiler).compile('<template>${$node}</template>', container.get(ViewResources));
+      }
+    } else {
+      this.nodeViewFactory = parent.nodeViewFactory;
+    }
+  }
+
+  @bindable
+  nodes: INode[];
+
+  @bindable
+  public theme: UxTreeViewTheme;
+  public themeChanged(newValue: UxTreeViewTheme) {
+    if (newValue !== null && !newValue.themeKey) {
+      newValue.themeKey = 'tree-view';
+    }
+
+    this.styleEngine.applyTheme(newValue, this.element);
+  }
+
+  nodeViewFactory: ViewFactory;
+  selectedNode: INode;
+
+  // this is populated by the HTML template
+  treeViews: UxTreeView[] = [];
+
+  toggleExpanded(n: INode, e: Event): boolean {
+    n.expanded = !n.expanded;
+    e.stopPropagation();
+    return false;
+  }
+
+  nodeClicked(n: INode) {
+    if (this.selectedNode) {
+      this.selectedNode.selected = false;
+    }
+    n.selected = true;
+    this.selectedNode = n;
+    this.element.dispatchEvent(new CustomEvent(
+      UxTreeView.NODE_SELECTED_EVENT,
+      { detail: { node: n }, bubbles: true })
+    );
+    return true;
+  }
+
+  childNodeSelected(n: INode) {
+    if (this.selectedNode && this.selectedNode !== n) {
+      this.selectedNode.selected = false;
+    }
+    this.selectedNode = n;
+  }
+
+  findPath(nodes: INode[], predicate: (node: INode) => boolean): number[] {
+    const path: number[] = [];
+    for (let i = 0; i < nodes.length; ++i) {
+      if (predicate(nodes[i])) {
+        return [i];
+      }
+      if (!nodes[i].children) {
+        continue;
+      }
+      const childPath = this.findPath(nodes[i].children!, predicate);
+      if (childPath.length) {
+        return [i, ...childPath];
+      }
+    }
+    return path;
+  }
+
+  expandPath(path: number[]) {
+    if (path.length === 1) {
+      this.nodeClicked(this.nodes[path[0]]);
+      this.element.querySelectorAll('.ux-tree-view--node')[path[0]].scrollIntoView();
+    } else {
+      this.nodes[path[0]].expanded = true;
+      // let Aurelia populate treeViews by queueing the task
+      this.taskQueue.queueTask(() => {
```

Comment:
if there's a search that returns 10 nodes in 10 different subtrees, doesn't this mean the last one will be focused?

Reply:
When a node found in the tree the path is returned immediately, so the first occurrence will be focused.

Reply:
I see. Should:
- it allow multiple paths return in `findPath`, say you have 10 departments, each department is a tree. and you want to search for a person name John in all those departments, limiting to a first department only is quite limited.
- this block
  ```ts
  if (path.length === 1) {
      this.nodeClicked(this.nodes[path[0]]);
  ```
  seems to be always click on the first path gradually, can you help explain why it's needed? Shouldn't that cause unwatned side effect - example: parent - child view where intermediate items would come into focus unnecessarily?

Reply:
The path array is a list of node indexes which need to be expanded. The last index is the one we're looking for

Reply:
as for multiple matches, didn't really think about that as the only requirement for me was to search by id which is unique in my case. Even if you have a name search, it is natural to focus the first occurrence

Reply:
yes, the last index will be focused, eventually. Though I think by faking click on the intermediate nodes, we will run into side effect issue, hypothetically like I mentioned before. Can we do the expand from code?

For multiple matches, I meant we are supposed to show in all subtrees, not just only the first one. Example:
```
company
  | department 1
      | - John Wick
  | department 2
      | - John Wicked
  | department 3
      | - John Wicker
  | department 4
      | - John Wickest
```
Searching for John should open the tree for 4 departments, but only focus on the first one.

Reply:
but the intermediate nodes are not clicked, only the last one

Reply:
as for multiple search - quite easy to do in flat mode, but a bit harder in hierarchy. I'd rather removed the API to let consumers do it.

Reply:
> but the intermediate nodes are not clicked, only the last one

Wait i see it. Yes it's is only clicking on the last one. I read the code wrong, sorry.

Reply:
> as for multiple search - quite easy to do in flat mode, but a bit harder in hierarchy. I'd rather remove the API to let consumers do it.

Ok lets get this in for now, I'll do an enhancement later when we need it